hack/tests/scaffolding/scaffold-memcached.go

```diff
@@ -147,13 +150,15 @@ func main() {
 			break
 		}
 	}
+
 	for lineNum, line := range memcachedTypesFileLines {
 		if strings.Contains(string(line), "type MemcachedStatus struct {") {
-			memcachedTypesFileLinesIntermediate := append(memcachedTypesFileLines[:lineNum+1], []byte("\tNodes []string `json:\"nodes\"`"))
+			memcachedTypesFileLinesIntermediate := append(memcachedTypesFileLines[:lineNum+1], []byte("\tNodes PodList `json:\"nodes\"`"))
```

Comment:
I agree with adding a check for the rule violations, but there's a much simpler fix that can be made here.

```diff
-                       memcachedTypesFileLinesIntermediate := append(memcachedTypesFileLines[:lineNum+1], []byte("\tNodes []string `json:\"nodes\"`"))
+                       memcachedTypesFileLinesIntermediate := append(memcachedTypesFileLines[:lineNum+1], []byte("\t// +listType=set"))
+                       memcachedTypesFileLinesIntermediate = append(memcachedTypesFileLinesIntermediate, []byte("\tNodes []string `json:\"nodes\"`"))
```

I think this PR can be reduced to just the above change and the section you added for "Checking API rule violations" in lines 182-186.

The changes to `test/test-framework` do not seem necessary to resolve the rule violations in the e2e test.

As far as I can tell, the `test/test-framework` code is only really used for the scorecard test. And as long as the scorecard test is passing, I don't see a need to make significant changes to it in the short term.

In the long term, I _do_ think we need to clean up our overall test structure and CI scripts since they are all over the place at the moment. But for that, we really need to have a comprehensive plan that lays out our desired end state and a plan to get from here to there. I'll plan to start putting a proposal together for that and circulate for feedback.

Reply:
HI @joelanford, 

It is passing with a code that should no longer work with. So, IMHO we should ensure that this mock data is updated with the current version always in order to ensure that the test is checking a valid impl.

Reply:
I didn't see `crd_test.go`'s dependency on `test/test-framework`. Good catch.

To try to isolate and reproduce this issue, I did the following:
1. Checked out `master`,
2. Ran `go test ./internal/scaffold/. -v -run TestCRDGoProject`, and it passed.
3. Added the `// +listType=set` annotation to `test/test-framework/pkg/apis/cache/v1alpha1/memcached_types.go`
4. Re-ran the test from step 2 with no other changes, and all tests still passed. 

So the `// +listType=set` annotation currently has no impact on `crd_test.go` or on the creation of the CRD, because I did not have to change the existing `crdCoExp` YAML string in `crd_test.go`.

Basically, what I'm still trying to re-iterate is that `test/test-framework` is an internal minimal set of project files that are required to execute SDK's tests. That is currently its only purpose. My opinion is that if and when we upgrade to a version of kubernetes dependencies that break our tests due to these rule violations, that will be a good signal to go and update `test/test-framework` AND the `_types.go` scaffolds we generate, if necessary.

At the end of the day, the biggest takeaway I have from this API rule violation investigation is that we need to make sure that we generate `_types.go` such that it would not result in rule violations. And I think the added check [here](https://github.com/operator-framework/operator-sdk/pull/2113/files#diff-80ec259a3981798e2b72eb84d4d72871R182-R186) accomplishes that goal.

So let's do this:
1. Keep the API rule violations check. That's a great addition to the Go e2e test. :+1:
2. Since we know about it now, let's go ahead and update the various memcached-operator implementations to use `// +listType=set` on `[]string` types. 
3. Keep the deepcopy.go changes, since that file is required for the subcommand test. Revert any other changes (e.g. openapi.go, Dockerfile, sanity tests, etc.), since none of those are required to make tests pass. We can debate these in a separate PR if you'd like.